internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,278 @@
+	setReplicas(s)
+
+	// Conditions
+
+	// Update the ScalingUp and ScalingDown condition, require the above setReplicas function.
+	setScalingUpCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound)
+	setScalingDownCondition(s.machineSet, s.machines, s.bootstrapObjectNotFound, s.infrastructureObjectNotFound, s.owningMachineDeployment, s.staleDeletingMachines)
+
+	// MachinesReady condition: aggregate the Machine's Ready condition
+	setMachinesReadyCondition(s.machineSet, s.machines)
+
+	// TODO MachinesUpToDate
+
+	// TODO Deleting
+}
+
+func setReplicas(s *scope) {
+	ms := s.machineSet
+	machines := s.machines
+	// Return early when machines is nil. It's not possible to calculate replica counters.
+	// Conditions may surface to be unknown in this case.
+	if machines == nil {
+		return
+	}
+
+	var readyReplicas, availableReplicas int32
+	for _, machine := range machines {
+		if meta.IsStatusConditionTrue(machine.GetV1Beta2Conditions(), clusterv1.MachineReadyV1Beta2Condition) {
+			readyReplicas++
+		}
+		if meta.IsStatusConditionTrue(machine.GetV1Beta2Conditions(), clusterv1.MachineAvailableV1Beta2Condition) {
+			availableReplicas++
+		}
+		if !machine.GetDeletionTimestamp().IsZero() && time.Since(machine.GetDeletionTimestamp().Time) >= time.Minute*30 {
+			s.staleDeletingMachines = append(s.staleDeletingMachines, machine.GetName())
+		}
+	}
+
+	if ms.Status.V1Beta2 == nil {
+		ms.Status.V1Beta2 = &clusterv1.MachineSetV1Beta2Status{}
+	}
+
+	ms.Status.Replicas = int32(len(machines))
+	ms.Status.V1Beta2.ReadyReplicas = ptr.To(readyReplicas)
+	ms.Status.V1Beta2.AvailableReplicas = ptr.To(availableReplicas)
+}
+
+func setScalingUpCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingUpErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
+	// This could e.g. be the case when using autoscaling with MachineSets.
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+	gotReplicas := ms.Status.Replicas
+
+	referencesMessage := calculateMissingReferencesMessage(bootstrapObjectNotFound, infrastructureObjectNotFound)
+
+	// Not scaling up or in deletion.
+	if gotReplicas >= desiredReplicas || !ms.DeletionTimestamp.IsZero() {
```

Comment:
unless there are specific use cases to address, I would not check deletion timestamp, but only rely on number of machines

Reply:
Changed to setting desiredReplicas to 0 instead on deletion.